crates/wasmtime/src/component/func/host.rs

```diff
@@ -330,3 +353,194 @@ macro_rules! impl_into_component_func {
 }
 
 for_each_function_signature!(impl_into_component_func);
+
+unsafe fn call_host_dynamic<T, F>(
+    Types { params, result }: &Types,
+    cx: *mut VMOpaqueContext,
+    mut flags: InstanceFlags,
+    memory: *mut VMMemoryDefinition,
+    realloc: *mut VMCallerCheckedAnyfunc,
+    string_encoding: StringEncoding,
+    storage: &mut [ValRaw],
+    closure: F,
+) -> Result<()>
+where
+    F: FnOnce(StoreContextMut<'_, T>, &[Val]) -> Result<Val>,
+{
+    let cx = VMComponentContext::from_opaque(cx);
+    let instance = (*cx).instance();
+    let mut cx = StoreContextMut::from_raw((*instance).store());
+
+    let options = Options::new(
+        cx.0.id(),
+        NonNull::new(memory),
+        NonNull::new(realloc),
+        string_encoding,
+    );
+
+    // Perform a dynamic check that this instance can indeed be left. Exiting
+    // the component is disallowed, for example, when the `realloc` function
+    // calls a canonical import.
+    if !flags.may_leave() {
+        bail!("cannot leave component instance");
+    }
+
+    let param_count = params.iter().map(|ty| ty.flatten_count()).sum::<usize>();
+    let result_count = result.flatten_count();
+
+    if param_count <= MAX_FLAT_PARAMS {
+        if result_count <= MAX_FLAT_RESULTS {
+            let ret = {
+                let iter = &mut storage.iter();
+                let args = params
+                    .iter()
+                    .map(|ty| Val::lift(ty, cx.0, &options, iter))
+                    .collect::<Result<Box<[_]>>>()?;
+                closure(cx.as_context_mut(), &args)?
+            };
+            result.check(&ret)?;
+            flags.set_may_leave(false);
+            let dst = mem::transmute::<_, &mut [MaybeUninit<ValRaw>]>(storage);
+            ret.lower(&mut cx, &options, &mut dst.iter_mut())?;
+        } else {
+            let iter = &mut storage.iter();
+            let args = params
+                .iter()
+                .map(|ty| Val::lift(ty, cx.0, &options, iter))
+                .collect::<Result<Box<[_]>>>()?;
```

Comment:
Above the 2x2 matrix of "params direct/indirect" x "results direct/indirect" is expressed with 4 different cases but the main reason for that is to get all the fiddly bits around the types to work out. Otherwise here though this may be more amenable to a straightforward "load the params, do the call, store the result" without having to specify 4 different cases (which leads to a fair bit of duplication between everything)